lib/grp.gi

```diff
@@ -483,27 +484,35 @@ local p, hom, reps, a, b, ap_bp, ab_p, H;
   reps := Filtered(reps, g -> not IsOne(g));
   reps := List(reps, g -> PreImagesRepresentative(hom, g));
 
-  for b in Image(hom) do
-    b := PreImagesRepresentative(hom, b);
-    for a in reps do
+  as := List(reps, a -> [a,a^p]);;
+  bs := List(Image(hom), function(b)
+      b := PreImagesRepresentative(hom, b);
+      return [b,b^p];
+      end);;
+
+  for b in bs do
+    bp := b[2]; b := b[1];
+    for a in as do
+      ap := a[2]; a := a[1];
       # if a and b commute the regularity condition automatically holds
-      if a*b = b*a then continue; fi;
+      ab := a*b;
+      if ab = b*a then continue; fi;
 
       # regularity is also automatic if a^p * b^p = (a*b)^p
-      ap_bp := a^p * b^p;
-      ab_p := (a*b)^p;
+      ap_bp := ap * bp;
+      ab_p := ab^p;
       if ap_bp = ab_p then continue; fi;
 
       # if the subgroup generated H by a and b is itself regular, we are also
-      # done. However we don't use recursion, here, as H may be equal to G;
-      # and also we have to be careful to not use too expensive code here.
-      # But a quick size check is certainly fine.
+      # done. However we don't use recursion here, as it is too expensive.
+      # we just check the direct definition, with a minor twist to avoid
+      # calling Agemo
+      g := ap_bp / ab_p;
+      h := Comm(a,b)^p;
+      if g = h or IsOne(g*h) then continue; fi;
       H := Subgroup(G, [a,b]);
-      if Size(H) <= p^p then continue; fi;
-
-      # finally the full check
-      H := DerivedSubgroup(H);
-      if not (ap_bp / ab_p) in Agemo(H, p) then
+      H := NormalClosure(H, [h]); # faster than Agemo(DerivedSubgroup(H), p);
+      if not g in H then
```

Comment:
Good. This new H should be much faster to compute, and if g is not in H, then G is surely not regular, but if G is regular, then this H is the same as the more expensive Agemo subgroup, and the check will always pass.

Reply:
I've now added a comment explaining what's going on, I hope this makes it much clearer. Apologies for not doing that before, I really should have :-(